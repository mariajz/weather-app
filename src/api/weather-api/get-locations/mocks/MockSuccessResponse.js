export const mockSuccessResponse = [
    {
        id: 1120673,
        name: 'Kochi',
        region: 'Kerala',
        country: 'India',
        lat: 9.97,
        lon: 76.23,
        url: 'kochi-kerala-india',
    },
    {
        id: 3125114,
        name: 'Kochi-Shi',
        region: 'Ko¬çchi',
        country: 'Japan',
        lat: 33.56,
        lon: 133.53,
        url: 'kochi-shi-kochi-japan',
    },
    {
        id: 1120674,
        name: 'Kochinda',
        region: 'Orissa',
        country: 'India',
        lat: 21.73,
        lon: 84.35,
        url: 'kochinda-orissa-india',
    },
];
