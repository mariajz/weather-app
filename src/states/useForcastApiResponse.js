import { atom, useAtom } from 'jotai';

const forcastApiResponseAtom = atom({});

const useForcastApiResponse = () => {
    const [response, setResponse] = useAtom(forcastApiResponseAtom);

    return {
        response,
        setResponse,
    };
};
export default useForcastApiResponse;
