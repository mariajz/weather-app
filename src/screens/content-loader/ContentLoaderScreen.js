import React, { useEffect } from 'react';
import { useNavigation } from '@react-navigation/native';
import Loader from '../../components/loader';
import useGetCurrentWeather from '../../hooks/useGetCurrentWeather';
import useForcastApiResponse from '../../states/useForcastApiResponse';

const ContentLoaderScreen = () => {
    const { handleFetchWeather } = useGetCurrentWeather();
    const { response } = useForcastApiResponse();
    const navigation = useNavigation();

    useEffect(() => {
        if (response && Object.keys(response).length === 0) {
            handleFetchWeather();
        }
    }, [handleFetchWeather, response]);

    useEffect(() => {
        if (response && Object.keys(response).length !== 0) {
            navigation.navigate('Dashboard', { data: response });
        }
    }, [navigation, response]);

    return <Loader />;
};

export default ContentLoaderScreen;
