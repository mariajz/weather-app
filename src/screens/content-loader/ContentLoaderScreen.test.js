import { render } from '@testing-library/react-native';
import React from 'react';
import { View as MockView } from 'react-native';
import { mockSuccessResponse } from '../../api/weather-api/current-weather/mocks/MockSuccessResponse';
import useForcastApiResponse from '../../states/useForcastApiResponse';
import ContentLoaderScreen from './ContentLoaderScreen';

jest.mock('../../components/loader', () => {
    return {
        __esModule: true,
        default: () => {
            return <MockView>Loader</MockView>;
        },
    };
});

const mockNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
    useNavigation: () => ({
        navigate: mockNavigate,
    }),
}));

jest.mock('../../states/useForcastApiResponse');

const mockHandleFetchWeather = jest.fn();
jest.mock('../../hooks/useGetCurrentWeather', () => () => ({
    handleFetchWeather: mockHandleFetchWeather,
}));

describe('ContentLoaderScreen', () => {
    beforeEach(() => {
        useForcastApiResponse.mockImplementation(() => ({
            response: mockSuccessResponse,
        }));
    });
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render ContentLoaderScreen', () => {
        const container = render(<ContentLoaderScreen />);

        expect(container).toMatchSnapshot();
    });

    it.each`
        response               | handleFetchWeatherCalledTimes | responseType
        ${mockSuccessResponse} | ${0}                          | ${'valid and non empty'}
        ${{}}                  | ${1}                          | ${'empty object'}
        ${undefined}           | ${0}                          | ${'undefined'}
    `(
        'should call handleFetchWeather $handleFetchWeatherCalledTimes times when response is $responseType',
        ({ response, handleFetchWeatherCalledTimes }) => {
            useForcastApiResponse.mockImplementation(() => ({
                response: response,
            }));

            render(<ContentLoaderScreen />);

            expect(mockHandleFetchWeather).toHaveBeenCalledTimes(
                handleFetchWeatherCalledTimes,
            );
        },
    );

    it('should navigate to dashboard when response is valid and non empty', async () => {
        useForcastApiResponse.mockImplementation(() => ({
            response: mockSuccessResponse,
        }));

        render(<ContentLoaderScreen />);

        expect(mockNavigate).toHaveBeenCalledWith('Dashboard', {
            data: mockSuccessResponse,
        });
    });
});
