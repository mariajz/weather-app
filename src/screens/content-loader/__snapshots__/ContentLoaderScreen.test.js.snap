// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentLoaderScreen should render ContentLoaderScreen 1`] = `
<View>
  Loader
</View>
`;
